### PR TITLE
FIX: liveness analysis now takes into account destructured struct parameters

### PR DESCRIPTION
**Pull request by Kobzol**
Fixes: https://github.com/intellij-rust/intellij-rust/issues/5109

### CODE REVIEW

**Comment by Undin** [comment]
@Kobzol FYI if PR makes some user-visible changes and/or you want that our QA check it (usually if a change is user-visible, you want it), add it into `To test` project

**Comment by Kobzol** [comment]
I think that I don't have permissions to add PRs to projects, there's no button on the Projects section in the right column.

**Comment by Undin** [comment]
@Kobzol it should be fixed now. Could you check it, please?

**Comment by Kobzol** [comment]
I just added the rest of my user-facing PRs to the `To test` project, so it is working :-) Thank you.

**Comment by ortem** [comment]
@Kobzol Changes in our control-flow analysis sometimes lead to false-positives in real projects. So I usually run `RsRealProjectAnalysisTest` (at least `Cargo` and `tokio` but more is better) before merging such PR, like [here](https://github.com/intellij-rust/intellij-rust/pull/5383).

Could you please run some tests from `RsRealProjectAnalysisTest` twice: before your changes and after them, and check there are no new false-positives related to `RsBorrowCheckerInspection` and `RsLivenessInspection`?

Note that `checkWarnings` should be `true` (see `RsRealProjectAnalysisTest.kt:74`). Also, probably you will have to comment `testAssert { implsAndTraits.distinct().size == implsAndTraits.size }` on `ImplLookup.kt:338` because it leads to early termination of some real project tests (BTW @vlad20012 can we deal with it somehow?)

File: src/test/kotlin/org/rust/ide/inspections/RsLivenessInspectionTest.kt [hunk 1]
```diff
@@ -489,9 +489,30 @@ class RsLivenessInspectionTest : RsInspectionsTestBase(RsLivenessInspection::cla
...
+            a: u32
+        }
+
+        fn foo() {
+            let S { a } = S { a: 0 };
+            a;
+        }
+    """)
+
+    fun `test destructured struct parameter`() = checkByText("""
```
Comment (ortem) [inline src/test/kotlin/org/rust/ide/inspections/RsLivenessInspectionTest.kt:509]: Please also ensure that destructured struct parameters work fine in move analysis and add such tests to `RsBorrowCheckerMovesTest`

**Comment by Kobzol** [comment]
I checked the real projects (except for `rustc`) and found no regressions. There are some false positives, but they were present even before this PR. And it fixed one false positive in `mysql_async`! :tada: 

This process should really be automated somehow :D Maybe we could write a Python/Bash script that takes two git revisions, runs the real projects tests on both of them and prints a diff of the emitted errors/warnings?

**Comment by ortem** [comment]
@Kobzol Great, thank you!

Having some kind of automation in this process would be very nice. We discussed it for a long time ago but still haven't had time to implement it.

bors r+

**Pull request merged**